mm2src/common/expirable_map.rs

```diff
@@ -47,6 +47,7 @@ impl<K: Eq + Hash, V> ExpirableMap<K, V> {
     /// If a value already exists for the given key, it will be updated and then
     /// the old one will be returned.
     pub fn insert(&mut self, k: K, v: V, exp: Duration) -> Option<V> {
+        self.clear_expired_entries();
```

Comment:
I might do some optimization to this https://github.com/KomodoPlatform/komodo-defi-framework/commit/62d26c7b59f2e47f6e50b7b1ecb6d66b8dcdf5ab. Will push the cherry-pick first then do the optimizations in another commit so that you can review them.

Reply:
> Only optimization I can find currently is to use retain in clear_expired_entries instead of for (exp, key) in keys_to_remove { loop

Nevermind, your current implementation is better :)

Reply:
Done here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232/commits/1c38da77a11be4eff685cf2e4f2f782d3021a94e

Reply:
`remove` function calls `clear_expired_entries` and I think it's not ideal. The entire implementation relies on manually managed expired entries. If we want to handle this automatically, we should do it in every function not just one. Currently, if I’ve already called `clear_expired_entries`, using the `remove` function adds unnecessary overhead. I’m also wondering if we can avoid `Clone`. I will look into that.

Reply:
Is `cloned` necessary?

Reply:
> Is `cloned` necessary?

Found a better alternative that doesn't use cloned
```rust
    pub fn clear_expired_entries_impl1(&mut self) {
        let now = Instant::now();

        while let Some((exp, key)) = self.expiries.pop_first() {
            if exp > now {
                self.expiries.insert(exp, key);
                break;
            }
            self.map.remove(&key);
        }
    }
```
<img width="838" alt="Screenshot 2024-09-30 at 11 35 44 AM" src="https://github.com/user-attachments/assets/74f0a546-1c25-4700-9dc1-1d48a4685922">

Reply:
> I’m also wondering if we can avoid Clone. I will look into that.

No need to check it yourself, I will check it now.

Reply:
Implemented this https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232#discussion_r1780664590 here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232/commits/da884fcb87a831716c524e9b317846d942b7289a

Reply:
I think we should either avoid calling `clear_expired_entries` anywhere in `ExpirableMap` or call it consistently throughout and update the documentation to reflect this (as it currently stands as manually managed expired entries).

The downside of calling `clear_expired_entries` within the binding functions is that when you insert/remove/get entries several times, `clear_expired_entries` will be triggered that many times within milliseconds. This is why it was designed in a manual way initially by leaving expiration management responsibility to the caller (e.g., I can clear the expired entries once and then insert/get/remove multiple keys).

Reply:
> I’m also wondering if we can avoid Clone. I will look into that.

Can't remove it without changing the current `ExpirableMap` structure, we can use `Arc` but it doesn't make sense, I think cloning once on insert is fine unless you have other ideas @onur-ozkan.

Reply:
> We don't really use get anywhere but I don't mind. The only thing is that it will have to be mutable func.
Also i think we can call clear_expired_entries on remove just fine. It's cheap and why wouldn't we.

No need to then, we can just check expiry on `get` and return `None` if expired.

Reply:
My concern is that we are semi-managing the expirations, which is confusing on use (we have to check which function are clearinging expired entries and which functions are not clearing it..). How about making `clear_expired_entries` private function, and using it internally everywhere in `ExpirableMap` implementation? It should be very clear that `ExpirableMap` is either controlling expired entries, or not doing it at all.

> Can't remove it without changing the current ExpirableMap structure, we can use Arc but it doesn't make sense, I think cloning once on insert is fine unless you have other ideas @onur-ozkan.

Yeah, it doesn't make sense to use `Arc`. What about using `Copy` instead of `Clone`? This should enforce callers to use `Copy`able (which are cheaper) types for keys.

Reply:
> Also i think we can call clear_expired_entries on remove just fine. It's cheap and why wouldn't we.

There is no need for `clear_expired_entries` on remove, the purpose of `clear_expired_entries` is to reduce the memory and we can do it on insert only.

Reply:
> Yeah, it doesn't make sense to use Arc. What about using Copy instead of Clone? This should enforce callers to use Copyable (which are cheaper) types for keys.

Will give copy a try.

Reply:
To summarize:

1 - Will make `clear_expired_entries` private
2 - Use it only on `insert`
3 - check expiry on `get`, not sure but we shouldn't return an expired entry in anyway.

Reply:
> > Also i think we can call clear_expired_entries on remove just fine. It's cheap and why wouldn't we.
> 
> There is no need for `clear_expired_entries` on remove, the purpose of `clear_expired_entries` is to reduce the memory and we can do it on insert only.

I agree for `remove`, but if we doing it for `insert`, we should do it for `get` too IMO. And if we do it for `insert` and `get`, `clear_expired_entries` can be private function.

Reply:
to do it on get, get will need a mutable self.

Reply:
Oooh, yeah.. That's not good.. I don't know what to suggest at this point :laughing:. As I said, a consistent interface is the concern for me. If it's not going to happen in every function (except remove as it doesn't make any sense), it should be handler from the caller side IMO as it's much clear to follow.

Reply:
Did this https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232#discussion_r1780828764 here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232/commits/5762d13c602d72c25ca2ae060c4402366235cebb

Reply:
Will try to replace clone with copy now.

Reply:
> Will try to replace clone with copy now.

Done here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232/commits/f2ef6d3a132d1ce52ea6ed3b085314a757c7bb6f

Reply:
> There is no need for clear_expired_entries on remove, the purpose of clear_expired_entries is to reduce the memory and we can do it on insert only.

Not just to reduce the memory, but also to enforce correctness. But if we are changing the API to tell the caller that `get` is the clearing version then it's OK.

Reply:
> What about using Copy instead of Clone? This should enforce callers to use Copyable (which are cheaper) types for keys.

Haha, i started with the same approach but chose clone eventually cuz tests used strings (non-copyable). Agree, this should be the way though.